narwhal/executor/src/subscriber.rs

```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
+            for worker in &workers {
+                let future = self.fetch_remote(stagger, worker.clone(), remaining_digests.clone());
+                futures.push(future.boxed());
+                // TODO: Make this a parameter, and also record workers / authorities that are down
+                // to request from them batches later.
+                stagger += Duration::from_millis(200);
+            }
+            let (remote_batches, _, _) = futures::future::select_all(futures).await;
+
+            for (remote_batch_digest, remote_batch) in remote_batches {
+                if remaining_digests.remove(&remote_batch_digest) {
+                    let batch_fetch_duration =
+                        remote_batch.metadata.created_at.elapsed().as_secs_f64();
+                    self.metrics
+                        .batch_execution_latency
+                        .observe(batch_fetch_duration);
+                    debug!(
+                        "Batch {:?} took {} seconds to be fetched for execution since creation",
+                        remote_batch_digest, batch_fetch_duration
+                    );
+
+                    fetched_batches.insert(remote_batch_digest, remote_batch);
+                }
+            }
+
+            if remaining_digests.is_empty() {
+                return fetched_batches;
+            } else {
+                error!("Failed to fetch all required batches from workers. Remaining batches: {remaining_digests:?}");
+            }
         }
-        let (batch, _, _) = futures::future::select_all(futures).await;
-        let batch_fetch_duration = batch.metadata.created_at.elapsed().as_secs_f64();
-        self.metrics
-            .batch_execution_latency
-            .observe(batch_fetch_duration);
-        debug!(
-            "Batch {:?} took {} seconds to be fetched for execution since creation",
-            batch.digest(),
-            batch_fetch_duration
-        );
-        batch
+        fetched_batches
     }
 
-    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
-    async fn try_fetch_locally(&self, digest: BatchDigest, worker_id: WorkerId) -> Option<Batch> {
+    #[instrument(level = "debug", skip_all, fields(digests = ? digests, worker_id = % worker_id))]
+    async fn try_fetch_locally(
+        &self,
+        digests: Vec<BatchDigest>,
+        worker_id: WorkerId,
+    ) -> HashMap<BatchDigest, Batch> {
         let _timer = self.metrics.subscriber_local_fetch_latency.start_timer();
         let worker = self.network.my_worker(&worker_id);
-        let payload = self.network.request_batch(digest, worker).await;
-        match payload {
-            Ok(Some(batch)) => {
-                debug!("Payload {} found locally", digest);
-                self.metrics.subscriber_local_hit.inc();
-                return Some(batch);
+        let request_batches_response = self.network.request_batches(digests, worker).await;
```

Comment:
what if we have more than 6MB of batches available locally?

Reply:
Whoops forgot to a continuation token to `RequestBatches`. Thanks!